### PR TITLE
refactor: use craft-application for "pack"

### LINKED ISSUE #316: Rebase "rockcraft pack" onto craft-application framework
### What needs to get done

The craft-application framework reduces repeated code from various *craft applications. Rockcraft should be rebased onto craft-application before we get a stable release.

### Why it needs to get done

Reduce repeated code throughout the *craft applications.

### PR DESCRIPTION
**Pull request by tigarmo**
This PR makes it so invocations of `rockcraft pack` get routed to the new craft-application-based "runner". This is a lot of code because "pack" basically needs everything that we do, but a lot of the code is copy/pasted from the now-legacy runner.

It is _strongly suggested_ to review per-commit

Fixes #316

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -44,7 +44,7 @@ module = ["craft_archives"]
 ignore_missing_imports = true
 
 [tool.pylint.messages_control]
-disable = "too-many-ancestors,too-few-public-methods,fixme,unspecified-encoding,use-implicit-booleaness-not-comparison,unnecessary-lambda-assignment,line-too-long"
+disable = "too-many-ancestors,too-few-public-methods,fixme,unspecified-encoding,use-implicit-booleaness-not-comparison,unnecessary-lambda-assignment,line-too-long,cyclic-import"
```
[the remark this answers is not in the document]
Comment (tigarmo) [reply pyproject.toml:47]: yeap. No idea why, the app itself is working fine and the other linters don't complain either

File: tests/unit/test_project.py [hunk 2]
```diff
@@ -592,3 +588,52 @@ def test_project_generate_metadata(yaml_loaded_data):
...
+                    build_for="amd64",
+                    base=BaseName(name="ubuntu", version="20.04"),
+                    platform_entry="amd64",
+                    build_for_variant=None,
+                ),
+            ],
+        ),
+    ],
+)
+def test_project_get_build_plan(yaml_loaded_data, platforms, expected_build_infos):
```
Comment (cmatsuoka) [inline tests/unit/test_project.py:636]: Maybe also test with platform names that don't match an architecture name?

File: rockcraft/services/image.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+from pathlib import Path
+from typing import cast
+
+from craft_application import AppMetadata, BaseService, ServiceFactory
+from craft_cli import emit
+
+from rockcraft import models, oci
+
+
+@dataclass
```
Comment (cmatsuoka) [inline rockcraft/services/image.py:31]: Can we set this as frozen?

File: rockcraft/services/lifecycle.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+class RockcraftLifecycleService(LifecycleService):
+    """Rockcraft-specific lifecycle service."""
+
+    @override
+    def setup(self) -> None:
+        """Configure extra args to the LifecycleManager."""
+        # pylint: disable=import-outside-toplevel
+        # This inner import is necessary to resolve a cyclic import
+        from rockcraft.services import RockcraftServiceFactory
```
Comment (cmatsuoka) [inline rockcraft/services/lifecycle.py:43]: Would it work if we import at toplevel, but protected with `TYPE_CHECKING`?

File: rockcraft/services/lifecycle.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            package_repositories=project.package_repositories or [],
+            project_name=project.name,
+            project_vars=project_vars,
+        )
+
+        super().setup()
+
+    @override
+    def run(self, step_name: str, part_names: list[str] | None = None) -> None:
+        """Overridden to configure package repositories."""
```
Comment (sergiusens) [inline rockcraft/services/lifecycle.py:65]: this feels more like a code comment, the override decorator takes care of the docstring (one of its uses)

File: tests/unit/test_project.py (hunk 2 above)
Comment (tigarmo) [reply tests/unit/test_project.py:653]: oh of course! Done

File: rockcraft/services/image.py (hunk 3 above)
Comment (tigarmo) [reply rockcraft/services/image.py]: done! I tried to set `RockcraftBuildInfo` also as frozen but then the base `BuildInfo` needs to be frozen too, which it probably right

File: rockcraft/services/lifecycle.py (hunk 5 above)
Comment (tigarmo) [reply rockcraft/services/lifecycle.py]: > the override decorator takes care of the docstring (one of its uses)

I thought so too but pydocstyle is complaining that the method doesn't have a docstring. I ended up duplicating the docstring and adding this as a code comment

File: rockcraft/services/lifecycle.py (hunk 4 above)
Comment (tigarmo) [reply rockcraft/services/lifecycle.py:43]: Interesting. If I do that, `pylint` complains that it doesn't know what `RockcraftServiceFactory` in the `cast()` call. However, if I do `cast("RockcraftServiceFactory", self._services)` it seems to work??

Comment (cmatsuoka) [reply rockcraft/services/lifecycle.py:43]: Maybe this solves the cyclic import warning too.

File: rockcraft/services/lifecycle.py (hunk 5 above)
Comment (sergiusens) [reply rockcraft/services/lifecycle.py]: there's a required rule in pydocstyle for this to work, in setup.cfg style:
```
[pydocstyle]
# D107 Missing docstring in __init__ (reason: documented in class docstring)
# D203 1 blank line required before class docstring (reason: pep257 default)
# D213 Multi-line docstring summary should start at the second line (reason: pep257 default)
ignore = D107, D203, D213
ignore_decorators = overrides
```

File: rockcraft/services/image.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                variant=None,  # TODO
+            )
+        else:
+            base_image, source_image = oci.Image.from_docker_registry(
+                project.base,
+                image_dir=image_dir,
+                arch=self._build_for,
+                variant=None,  # TODO
+            )
+        emit.progress(f"Retrieved base {project.base} for {build_for}", permanent=True)
```
Comment (sergiusens) [inline rockcraft/services/image.py:84]: maybe a chance to make these not permanent

File: rockcraft/services/image.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+        emit.progress(f"Extracting {base_image.image_name}")
+        rootfs = base_image.extract_to(bundle_dir)
+        emit.progress(f"Extracted {base_image.image_name}", permanent=True)
```
Comment (sergiusens) [inline rockcraft/services/image.py:88]: maybe a chance to make these not permanent

File: rockcraft/services/service_factory.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+
+@dataclass
+class RockcraftServiceFactory(ServiceFactory):
+    """Rockcraft-specific Service Factory."""
+
+    # pylint: disable=invalid-name
+
+    ImageClass: type[services.RockcraftImageService] = services.RockcraftImageService
+    PackageClass: type[base_services.PackageService] = services.RockcraftPackageService
```
Comment (sergiusens) [inline rockcraft/services/service_factory.py:40]: maybe add a comment that this is an override?

**Review by sergiusens — APPROVED** [review]

**Comment by sergiusens** [comment]
Not an expert on the ServiceFactory, but this looks good. Great work

File: rockcraft/services/lifecycle.py (hunk 4 above)
Comment (cmatsuoka) [reply rockcraft/services/lifecycle.py:43]: Just checked, it doesn´t.

**Review by cmatsuoka — APPROVED** [review]
+1. Not a big fan of suppressing cyclic import warnings but this seems to be bogus.

**Pull request merged**